plugins/builds/index.js

```diff
@@ -515,6 +518,83 @@ async function getParentBuildStatus({ newBuild, joinListNames, pipelineId, build
     return { hasFailure, done };
 }
 
+/**
+ * [handleStage description]
+ * @param  {[type]} config               [description]
+ * @return {[type]}        [description]
+ */
+async function handleStage(config) {
+    const {
+        done,
+        hasFailure,
+        jobName,
+        newBuild,
+        jobFactory,
+        buildFactory,
+        pipelineId,
+        username,
+        scmContext,
+        event,
+        stage,
+        stageBuildFactory
+    } = config;
+
+    if (!stage || !done || !['CREATED', null, undefined].includes(newBuild.status)) {
+        return newBuild;
+    }
+
+    const stageSetupName = getFullStageJobName({ stageName: stage.name, jobName: 'setup' });
+    const stageTeardownName = getFullStageJobName({ stageName: stage.name, jobName: 'teardown' });
+
+    // Create/run stage teardown if stage job failure and no other running jobs in stage
+    if (hasFailure) {
+        // New build is stage teardown job
+        if (jobName === stageTeardownName) {
+            return newBuild;
+        }
+        // Remove current job
+        logger.info(
+            `Failure occurred in upstream job, removing new build - build:${newBuild.id} pipeline:${pipelineId}-${jobName} event:${newBuild.eventId} `
+        );
+        await newBuild.remove();
+
+        // Check if stage teardown build already exists
+        const stageTeardownJob = await jobFactory.list({ params: { pipelineId, name: stageTeardownName } });
+
+        if (stageTeardownJob) {
+            const existingStageTeardownBuild = await buildFactory.list({
+                params: { eventId: event.id, jobId: stageTeardownJob[0].id }
+            });
+
+            if (existingStageTeardownBuild && existingStageTeardownBuild[0]) {
+                return existingStageTeardownBuild[0];
+            }
+        }
+
+        // Doesn't exist, create stage teardown job and return as next job
+        const stageTeardownBuild = await createInternalBuild({
+            jobFactory,
+            buildFactory,
+            pipelineId,
+            jobName: stageTeardownName,
+            username,
+            scmContext,
+            event, // this is the parentBuild for the next build
+            baseBranch: event.baseBranch || null,
+            start: false
+        });
+
+        return stageTeardownBuild;
+    }
```

Comment:
If a stage job fails, `teardown` build would be created via `removeJoinBuilds` logic.
Why do we need to create `teardown` build here?